Add hier config nested compliance rule feature

Dav-C opened the pull request:
<!--
    Thank you for your interest in contributing to Golden Config! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->

# Closes: #978 

## What's Changed

- Added feature utilizing hier config for compliance rule matching nested configuration lines.
- Updated Tests
- Updated Docs

## To Do

<!--
    Please feel free to update todos to keep track of your own notes for WIP PRs.
-->
- [x] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

Commit:
nested compliance in progress

Commit:
update comments

Commit:
update section config eact match argument

Commit:
update nested compliance

Commit:
add hier config compliance option

Commit:
update hier config compliance

Commit:
update hier config compliance

Commit:
add config type check, rename os variable

Commit:
Merge branch 'develop' into dcates-nested-compliance

Commit:
remove dunder based nested compliance

Commit:
Merge branch 'develop' into dcates-nested-compliance

Commit:
add tests and docs

Commit:
Merge branch 'develop' into dcates-nested-compliance

Commit:
update nested compliance for hier config v3

Commit:
Add hier config nested compliance rule feature

Commit:
ruff

Commit:
add changelog fragment

itdependsnetworks commented on docs/user/app_feature_compliance_hierarchical.md:
```diff
@@ -0,0 +1,298 @@
+# Hierarchical Configuration Compliance
```
We should come up with a word that will resonate with the network engineer vs the technical implementation, such as partial match or filter configs.

steve-hodgkiss replied:
If filtered doesn’t work for you, then how about “Advanced configuration compliance”

jeffkala replied:
My vote is `Limited Configuration Compliance`

itdependsnetworks commented on nautobot_golden_config/nornir_plays/config_compliance.py:
```diff
@@ -16,6 +21,7 @@
 from nornir.core.plugins.inventory import InventoryPluginRegister
 from nornir.core.task import Result, Task
 from nornir_nautobot.exceptions import NornirNautobotException
+from pydantic import TypeAdapter
```
This is not a **direct** dependency in GC, we should consider if we want to make it one or not.

Dav-C replied:
pydantic is required by hier-config, so I assume it will always be available.

glennmatthews replied:
in theory hier-config could do a new release that no longer requires pydantic? Unlikely perhaps, but better to be explicit with regard to dependencies.

Dav-C replied:
Better safe than sorry, I added the dependency and kept it inline with what hier-config's requirements

itdependsnetworks commented on nautobot_golden_config/nornir_plays/config_compliance.py:
```diff
@@ -171,8 +177,17 @@ def run_compliance(  # pylint: disable=too-many-arguments,too-many-locals
+        # Check for hier config compliance rule
+        section_type = rule["section"][0]
+        if "# hier_config" in section_type:
+            hier_config_rule = True
+        if hier_config_rule:
```
I would just remove these two lines.

itdependsnetworks commented on docs/user/app_feature_compliance_hierarchical.md:
```diff
@@ -0,0 +1,298 @@
+
+Hierarchical compliance is particularly useful when:
+
+- You need to check compliance for complex nested configurations (e.g., BGP neighbors, interface configurations)
+- Standard CLI compliance matching is too broad or doesn't capture the specific configuration elements you want to validate
+- You need to apply compliance rules based on configuration hierarchy and relationships rather than simple line matching
+- You want to leverage the advanced parsing capabilities of the `hier_config` library
```
In general, let's keep to the use case vs the technology.

Dav-C replied:
The docs have been simplified and the focus is on the examples. I'm open to suggestions here, there probably could be more/better examples included.

itdependsnetworks commented on docs/user/app_feature_compliance_hierarchical.md:
```diff
@@ -0,0 +1,298 @@
+
+To create a hierarchical compliance rule, you must include a special comment marker in the **Config to Match** field of your compliance rule:
+
+- For hier_config v2 syntax: `# hier_config_v2`
```
Let's simplify this documentation after the removal of v3, in general, we shouldn't be re-documenting what hier_config does, but instead provide a few examples, and reference hier_config.

Commit:
remove hier config v2 syntax option

Commit:
update docs

itdependsnetworks commented on docs/user/app_feature_compliancefiltered.md:
```diff
@@ -0,0 +1,215 @@
+2. **Platform Compatibility**: Your device platform must be supported by the `hier_config` library
+3. **Network Driver Mapping**: The platform must have a valid `hier_config` mapping in its `network_driver_mappings`
+
+### Repository Settings
+
+The same repository settings required for standard compliance apply:
+
+- Backup repository for storing device configurations
+- Intended configuration repository
+- Proper `backup_path_template` and `intended_path_template` configuration
```
I think we can remove.

Dav-C replied:
removed

jeffkala commented:
Think my biggest concern is the actual and intended views are `filtered` too. I fully agree a filtered compliance has tons of value, but that doesn't mean the intended/actual should be filtered as well. (IMHO). I think the below should happen.

1. Actual should show as if the old "search" is being used. "interface" is the root/parent and it shows the entire interface section with all config.
2. Intended ^^ same as above
3. The actual compliance result should be filtered down to the hier_config rule definition.

steve-hodgkiss commented:
I think filtered is the whole point TBH. Compliance rules as they stand today are currently filtered if you think about it from a certain point of view - you’re matching on the parent command rather than displaying the whole intended in one screen. For example my intended config can have loads of lines which don’t have a compliance rule set and the app is happy with all the rules green.

jeffkala commented:
> I think filtered is the whole point TBH. Compliance rules as they stand today are currently filtered if you think about it from a certain point of view - you’re matching on the parent command rather than displaying the whole intended in one screen. For example my intended config can have loads of lines which don’t have a compliance rule set and the app is happy with all the rules green.

I understand your viewpoint; however, it shifts many other users use cases. Many people use the other config plan types for example (intended) which pulls from the compliance result. This now means its a filtered intended config-set which isn't what most users want.  I think the key components are to make this work alongside all the apps features/functionalities. Where the intended and actual are still showing "everything" that matches the root/main parent, and then the missing/extra/compliance result filtered down.

jeffkala commented:
## Current PR Does

Actual:

```
router bgp 100
    secret abcd
```

Intended:

```
router bgp 100
    secret defg
```

Compliance:

```
router bgp 100
    secret abcd
```

## Follows Existing pattern

Actual:

```
router bgp 100
    neighbor 1.1.1.1 remote-as 200
    secret abcd
    prefix-list JEFF
```

Intended:

```
router bgp 100
    neighbor 1.1.1.1 remote-as 200
    secret defg
    prefix-list JEFF
```


Compliance:

```
router bgp 100
    secret abcd
```

jeffkala committed:
Merge branch 'develop' into dcates-nested-compliance

mzbroch commented:
I had the opportunity to review this new feature and wanted to share some thoughts from an outside perspective:
- It's important that we remain consistent at every step.
- This means we should look at the solution as a whole, not just in terms of a new feature — everything should feel unified.
- From my point of view, HIER-CLI is simply a specific type of configuration. This implies that (a) matching, (b) processing, and (c) reporting behave in a particular way once I decide that my ComplianceRule is of type HIER-CLI.
- If we decide to introduce a header like # hier-cli, we should apply it consistently — across all types. If that's not feasible, perhaps we should reconsider whether the header is needed at all, especially if we can achieve the same clarity by introducing a new, dedicated type like "hier-cli".
- There’s no real need to include more "intended config" for a rule than what’s being matched and reported. Here’s why:
- We already follow a clear pattern — we match a section of the config and report that specific part. We don’t show the entire config above the matched snippet. Even when we have the full config available, we still only focus on the relevant fragment.
- Introducing names like "Filtered configuration compliance" or "Advanced CLI Compliance" might sound overly complex or too heavy. Couldn’t we simply stick with the term HIER-CLI, which clearly conveys the context — i.e. hierarchical configs with a hierarchical processing engine?

nautobot-bot committed:
Cookie updated by NetworkToCode Cookie Drift Manager Tool (#1048)

* Cookie updated by NetworkToCode Cookie Drift Manager Tool

Template:

```
{
    "template": "https://github.com/nautobot/cookiecutter-nautobot-app.git",
    "dir": "nautobot-app",
    "ref": "nautobot-app-v2.7.2",
    "path": null
}
```

Cookie:

```
{
    "remote": "https://github.com/nautobot/nautobot-app-golden-config.git",
    "path": "/tmp/tmpa4p9h4k6/nautobot-app-golden-config",
    "repository_path": "/tmp/tmpa4p9h4k6/nautobot-app-golden-config",
    "dir": "",
    "branch_prefix": "drift-manager/ltm-2.4",
    "context": {
        "codeowner_github_usernames": "@itdependsnetworks @jeffkala @nkallergis",
        "full_name": "Network to Code, LLC",
        "email": "opensource@networktocode.com",
        "github_org": "nautobot",
        "app_name": "nautobot_golden_config",
        "verbose_name": "Golden Config",
        "app_slug": "nautobot-golden-config",
        "project_slug": "nautobot-app-golden-config",
        "repo_url": "https://github.com/nautobot/nautobot-app-golden-config",
        "base_url": "golden-config",
        "camel_name": "NautobotGoldenConfig",
        "project_short_description": "An app for configuration on nautobot",
        "model_class_name": "ComplianceFeature",
        "open_source_license": "Apache-2.0",
        "docs_base_url": "https://docs.nautobot.com",
        "docs_app_url": "https://docs.nautobot.com/projects/golden-config/en/latest",
        "_extensions": [
            "local_extensions.camel_case_to_kebab",
            "local_extensions.camel_case_to_words",
            "local_extensions.NautobotVersions"
        ],
        "_template": "https://github.com/nautobot/cookiecutter-nautobot-app.git",
        "_output_dir": "/tmp/tmpa4p9h4k6",
        "_repo_dir": "/github/home/.cookiecutters/cookiecutter-nautobot-app/nautobot-app",
        "_checkout": "nautobot-app-v2.7.2"
    },
    "drift_managed_branch": "ltm-2.4",
    "remote_name": "origin",
    "pull_request_strategy": "PullRequestStrategy.UPDATE_OR_CREATE",
    "post_actions": [],
    "baked_commit_ref": "3e6d37c915ae6c6cde371e2065fffb07780a54f0",
    "draft": false
}
```

CLI Arguments:

```
{
    "cookie_dir": "",
    "input": false,
    "json_filename": "",
    "output_dir": "",
    "push": true,
    "template": "",
    "template_dir": "",
    "template_ref": "nautobot-app-v2.7.2",
    "pull_request": "update-or-create",
    "post_action": [],
    "disable_post_actions": true,
    "draft": null,
    "drift_managed_branch": "ltm-2.4"
}
```

* Apply suggestion from @gsnider2195

---------

Co-authored-by: bakebot <info@networktocode.com>
Co-authored-by: Gary Snider <75227981+gsnider2195@users.noreply.github.com>

gsnider2195 committed:
release v2.6.0 (#1049)

mzbroch committed:
Improve logic in get_repo_types_for_job (#1052)

jeffkala committed:
Merge pull request #1056 from nautobot/mzb/ltm-2.4/1051

[LTM 2.4] Improve logic in get_repo_types_for_job (#1052)

jeffkala committed:
Release v2.6.1

jeffkala committed:
Release v2.6.1

jeffkala committed:
Merge pull request #1057 from nautobot/release-2.6.1

Release 2.6.1

yanshcodilime committed:
Backported UI fixes from develop to ltm-2.4 (#1053)

* Added fixes

* Added changelog

* Added command to generate more data

* Ruff fix

joewesch committed:
Fixed Hier Config Remediation Options

yanshcodilime committed:
Updated django-debug-toolbar version (#1069)

* Updated django-debug-toolbar version

* Changelog

* Apply suggestion from @smk4664

* Update poetry.lock

---------

Co-authored-by: Stephen Kiely <stephenkiely87@gmail.com>

joewesch committed:
Changed compliance rule match_config to preserve leading spaces

joewesch committed:
Merge pull request #1067 from nautobot/u/joewesch-fix-remediation-options-ltm

[LTM 2.4 Backport] Fixed Hier Config Remediation Options

joewesch committed:
Merge pull request #1075 from nautobot/u/joewesch-match-config-leading-spaces-ltm

[LTM 2.4 Backport] Changed compliance rule match_config to preserve leading spaces

Defiantearth committed:
Release v2.6.2

itdependsnetworks committed:
Merge pull request #1076 from nautobot/release-2.6.2

Release v2.6.2

joewesch committed:
Fixed isnull filter getting added to extra filters (#1096)

joewesch committed:
Release v2.6.3

joewesch committed:
Merge pull request #1099 from nautobot/release-2.6.3

Release v2.6.3

joewesch committed:
Fixed a race condition when pushing to remote git repos (#1108)

joewesch committed:
prepare release v2.6.4

joewesch committed:
Merge pull request #1111 from nautobot/release/2.6.4

Release v2.6.4

Commit:
Merge remote-tracking branch 'upstream/ltm-2.4' into dcates-fix-merge-confilcts

Commit:
Merge branch 'dcates-fix-merge-confilcts' into dcates-nested-compliance

Commit:
remove duplicate function

Commit:
Merge remote-tracking branch 'upstream/develop' into dcates-fix-merge-confilcts

Commit:
update docs and config compliance

Commit:
Merge branch 'dcates-fix-merge-confilcts' into dcates-nested-compliance

Commit:
update documentation

Dav-C commented on docs/user/app_feature_compliancefiltered.md:
```diff
@@ -0,0 +1,215 @@
+
+Filtered configuration compliance requires:
+
+1. **CLI Configuration Type**: Only `CLI` configuration types are supported for hierarchical compliance
```
I added a link to the cli compliance feature within these docs (app_feature_compliancecli.md). Is that sufficient?

Commit:
add pydantic dependency

Commit:
update readme and docs

Commit:
update tests

itdependsnetworks commented on docs/user/app_feature_compliancefiltered.md:
````diff
@@ -0,0 +1,215 @@
+ permit ip 192.168.1.0 0.0.0.255 any
+ deny ip any any log
+```
+
+## Fallback Behavior for Empty Intended Configuration
+
+In some cases, the intended configuration may not contain any elements that match the compliance rules, resulting in an empty intended configuration text. To handle this scenario, the system includes a fallback mechanism:
+
+**Interface Fallback**: When the intended configuration text is empty, the system automatically looks for top-level interface configurations in the running configuration and uses them as the intended configuration baseline.
+
+This fallback behavior:
+
+1. **Triggers** when `intended_text` is empty after tag-based filtering
+3. **Filters** for lines that start with "interface"
+4. **Uses** these interface declarations as the intended configuration for comparison
+
+**Example Scenario**:
+- Your hierarchical rule targets specific VLAN configurations
+- The intended configuration template doesn't include those VLANs
+- The running configuration has interface declarations
+- The system uses the interface lines from running config as the baseline
+
+This ensures that remediation will not remove interfaces when intended configuration is empty for an interface.
+
````
I still feel similar, I don't fully follow the solution (examples would help), and it seems like more complication then required in a day 1 solution.

itdependsnetworks commented on nautobot_golden_config/nornir_plays/config_compliance.py:
```diff
@@ -194,6 +205,98 @@ def run_compliance(  # pylint: disable=too-many-arguments,too-many-locals
+        raise NornirNautobotException(error_msg) from e
+
+    # Apply tags to the running and generated configs
+    for tag_rule in v3_tags:
+        for child in running_config.get_children_deep(tag_rule.match_rules):
+            child.tags_add(tag_rule.apply_tags)
+
+    for tag_rule in v3_tags:
+        for child in generated_config.get_children_deep(tag_rule.match_rules):
+            child.tags_add(tag_rule.apply_tags)
```
Shouldn't this be applied the same on running and generated? When does `get_children_deep` get applied to a running but not generated config?
